File: packages/core/src/tenderiq_core/findings.py
"""Bulgu sözleşme tipleri (Sprint 2.2) — ajan şemaları ile ORM'nin ortak enum'ları.

``parsing.types`` deseniyle aynı: enum'lar tek kaynaktır ve bağımlılık yönü
gereği (models ← agents, B.2) nötr bir modülde yaşar — ORM modelleri ``agents``
paketini import etmeden bu tipleri kullanır (döngüsel import kırılır). Ajan
şemaları (``agents.schemas``) ve grounding (``agents.grounding``) bu tipleri
yeniden dışa aktarır; ajan sözleşmesi ile DB şeması ayrışamaz.
"""

from __future__ import annotations

from enum import StrEnum


class RequirementKind(StrEnum):
    """Gereksinim tipi (§8.1 ``Requirement``)."""

    TECHNICAL = "technical"  # teknik şartlar (ürün/hizmet nitelikleri, SLA...)
    ADMINISTRATIVE = "administrative"  # idari şartlar (yeterlik, usul, süreler...)
    FINANCIAL = "financial"  # mali/ekonomik şartlar (ciro, bilanço, teminat oranı...)


class DeliverableKind(StrEnum):
    """Sunulacak belge tipi (§8.1 ``Deliverable``)."""

    DOCUMENT = "document"  # idari/teknik belge (iş deneyim belgesi, taahhütname...)
    CERTIFICATE = "certificate"  # sertifika/kalite belgesi (ISO, TSE, CE...)
    GUARANTEE = "guarantee"  # teminat (geçici/kesin teminat mektubu...)
    OTHER = "other"


class RiskSeverity(StrEnum):
    """Risk maddesinin önem derecesi (Sprint 2.3, §8.2 ``RiskFlag``)."""

    LOW = "low"  # olağan/standart; farkındalık yeterli
    MEDIUM = "medium"  # dikkat gerektirir (ör. yüksek gecikme cezası oranı)
    HIGH = "high"  # tekliften önce mutlaka değerlendirilmeli (ör. sınırsız sorumluluk)


class RiskCategory(StrEnum):
    """Risk maddesinin türü (TR ihale alanı; getirim şablonlarıyla hizalı)."""

    PENALTY = "penalty"  # cezai şart / gecikme cezası
    TERMINATION = "termination"  # sözleşmenin feshi / yasaklılık halleri
    WARRANTY = "warranty"  # garanti / bakım yükümlülükleri
    PAYMENT = "payment"  # ödeme koşulları / fiyat farkı verilmemesi
    OTHER = "other"  # yukarıdakilere uymayan olağandışı/riskli madde


class TimelineKind(StrEnum):
    """Takvim öğesinin türü (Sprint 2.3, §8.2 ``TimelineEvent``)."""

    TENDER_DATE = "tender_date"  # ihale (açılış) tarihi
    BID_DEADLINE = "bid_deadline"  # son teklif verme tarihi/saati
    DELIVERY = "delivery"  # işin süresi / teslim programı
    WARRANTY = "warranty"  # garanti/bakım süresi
    OTHER = "other"  # diğer tarih/süre (ör. teklif geçerlilik süresi)


class ComplianceStatus(StrEnum):
    """Gereksinimin firma yetkinlik profiline göre karşılanma durumu (§6.7)."""

    MET = "met"  # profil gereksinimi tam karşılıyor
    PARTIAL = "partial"  # kısmen karşılanıyor / belirsiz — insan incelemesi gerek
    UNMET = "unmet"  # profil gereksinimi karşılamıyor (kapanması gereken boşluk)


class GroundingResolution(StrEnum):
    """Bulgunun kaynağa bağlanma düzeyi (§6.9, ADR-0006)."""

    ELEMENT = "element"  # tek ParsedElement'te doğrulandı (en keskin bağ)
    CHUNK = "chunk"  # chunk'ta doğrulandı; öğe sınırı aşan alıntı → aralık başı
    UNGROUNDED = "ungrounded"  # kaynakta doğrulanamadı — düşük güven, gösterilmez


class ReviewStatus(StrEnum):
    """Bulgunun insan-döngüde inceleme durumu (Sprint 3.2, §4.3).

    Çıkarım sonrası her bulgu PENDING doğar; insan onaylar, düzeltir (içerik
    değişikliği) veya reddeder. EDITED onaylı sayılır (düzeltilmiş hâliyle);
    REJECTED bulgular export'a girmez. Yeniden çıkarım (delete+insert) inceleme
    durumunu bilinçli sıfırlar — yeni çıkarım yeni inceleme gerektirir.
    """

    PENDING = "pending"  # onay bekliyor (varsayılan)
    APPROVED = "approved"  # insan onayladı
    EDITED = "edited"  # insan içeriği düzeltti (onaylı sayılır)
    REJECTED = "rejected"  # insan reddetti (hatalı/geçersiz bulgu)


class FindingKind(StrEnum):
    """Bulgu koleksiyonlarının ortak adresleme anahtarı (yorum/geçmiş uçları).

    Değerler AuditLog ``resource_type`` alanında ve API yollarında kullanılır;
    inceleme UI'sindeki beş sekmeyle birebir eşleşir.
    """

    REQUIREMENT = "requirement"
    DELIVERABLE = "deliverable"
    RISK = "risk"
    TIMELINE = "timeline"
    COMPLIANCE = "compliance"

File: apps/web/src/lib/findings.ts
import type { components, paths } from "@tenderiq/api-client";

import type { StatusTone } from "@/components/status-pill";

/** Üretilen istemciden bulgu yanıt tipleri (B.5 sözleşmesi — elle tip yazılmaz). */
type Ok<P extends keyof paths> = paths[P] extends {
  get: { responses: { 200: { content: { "application/json": infer T } } } };
}
  ? T
  : never;

export type RequirementFinding = Ok<"/api/v1/tenders/{tender_id}/requirements">[number];
export type DeliverableFinding = Ok<"/api/v1/tenders/{tender_id}/deliverables">[number];
export type RiskFinding = Ok<"/api/v1/tenders/{tender_id}/risks">[number];
export type TimelineFinding = Ok<"/api/v1/tenders/{tender_id}/timeline">[number];
export type ComplianceFinding = Ok<"/api/v1/tenders/{tender_id}/compliance">[number];
export type FindingSource = RequirementFinding["source"];

export type AnyFinding =
  | RequirementFinding
  | DeliverableFinding
  | RiskFinding
  | TimelineFinding
  | ComplianceFinding;

export type FindingCategory =
  | "requirements"
  | "deliverables"
  | "risks"
  | "timeline"
  | "compliance";

export const CATEGORY_LABELS: Record<FindingCategory, string> = {
  requirements: "Gereksinimler",
  deliverables: "Belgeler",
  risks: "Riskler",
  timeline: "Takvim",
  compliance: "Uygunluk",
};

/** İnceleme uçlarının (yorum/geçmiş/bulk) kullandığı tekil bulgu türü anahtarı. */
export type FindingKind = components["schemas"]["FindingKind"];
export type ReviewStatus = components["schemas"]["ReviewStatus"];
export type ReviewAction = components["schemas"]["ReviewAction"];

export const CATEGORY_TO_KIND: Record<FindingCategory, FindingKind> = {
  requirements: "requirement",
  deliverables: "deliverable",
  risks: "risk",
  timeline: "timeline",
  compliance: "compliance",
};

/** İnceleme durumu rozetleri (§7.8 tonları): üç durum + red (Sprint 3.2). */
export const REVIEW_STATUS: Record<ReviewStatus, { label: string; tone: StatusTone }> = {
  pending: { label: "Onay bekliyor", tone: "neutral" },
  approved: { label: "Onaylandı", tone: "success" },
  edited: { label: "Düzeltildi", tone: "info" },
  rejected: { label: "Reddedildi", tone: "danger" },
};

export const REQUIREMENT_KIND_LABELS: Record<string, string> = {
  technical: "Teknik",
  administrative: "İdari",
  financial: "Mali",
};

export const DELIVERABLE_KIND_LABELS: Record<string, string> = {
  document: "Belge",
  certificate: "Sertifika",
  guarantee: "Teminat",
  other: "Diğer",
};

export const RISK_SEVERITY: Record<string, { label: string; tone: StatusTone }> = {
  high: { label: "Yüksek", tone: "danger" },
  medium: { label: "Orta", tone: "warning" },
  low: { label: "Düşük", tone: "neutral" },
};

export const RISK_CATEGORY_LABELS: Record<string, string> = {
  penalty: "Cezai şart",
  termination: "Fesih",
  warranty: "Garanti",
  payment: "Ödeme",
  other: "Diğer",
};

export const TIMELINE_KIND_LABELS: Record<string, string> = {
  tender_date: "İhale tarihi",
  bid_deadline: "Son teklif",
  delivery: "Teslim/süre",
  warranty: "Garanti süresi",
  other: "Diğer",
};

export const COMPLIANCE_STATUS: Record<string, { label: string; tone: StatusTone }> = {
  met: { label: "Karşılanıyor", tone: "success" },
  partial: { label: "Kısmen", tone: "warning" },
  unmet: { label: "Karşılanmıyor", tone: "danger" },
};

/** Kaynak referansı metni: `s. 12 · §2.3 Teknik Şartlar` (mono ile dizilir). */
export function sourceRef(source: FindingSource): string {
  const section = source.section?.trim();
  return section != null && section !== "" ? `s. ${source.page} · ${section}` : `s. ${source.page}`;
}

/** PDF vurgu katmanının çizeceği kutu; konumsuz formatlarda (DOCX/XLSX) null. */
export function sourceBbox(
  source: FindingSource,
): { x0: number; y0: number; x1: number; y1: number } | null {
  if (
    source.bbox_x0 == null ||
    source.bbox_y0 == null ||
    source.bbox_x1 == null ||
    source.bbox_y1 == null
  ) {
    return null;
  }
  return { x0: source.bbox_x0, y0: source.bbox_y0, x1: source.bbox_x1, y1: source.bbox_y1 };
}
